mlrun/api/db/sqldb/session.py

```diff
@@ -12,59 +12,60 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+import typing
+
 from sqlalchemy import create_engine
 from sqlalchemy.engine import Engine
 from sqlalchemy.orm import Session
 from sqlalchemy.orm import sessionmaker as SessionMaker
 
 from mlrun.config import config
 
-engine: Engine = None
-_session_maker: SessionMaker = None
-
 
-# doing lazy load to allow tests to initialize the engine
-def get_engine() -> Engine:
-    global engine
-    if engine is None:
-        _init_engine()
-    return engine
+def create_session(dsn=None) -> Session:
+    session_maker = DBEngine(dsn=dsn)._get_session_maker()
```

Comment:
Also, and a bigger problem - you're creating a new `DBEngine` every time this is called. This is not what we need. You can still wrap the functionality in a `DBEngine` class that will hold the engine itself and the session-maker, but you want to cache these object globally. So, instead of holding a global `Engine`, you'll need to move your `_engines` to the global scope, and not in the `DBEngine` class.
Then in this method you need to check if you already have an engine for the given `dsn` and if not, generate it and provide the session-maker from it.

Reply:
I’ve brought `session.py` back to the original structure (removed the `DBEngine` class) and adjusted the global dictionaries. In addition, the `dsn` param is now passing through all of the functions within that file. This is the simplest and most backward compatibility approach.  Now instead of creating an instance of the `DBEngine` class, we can just call `create_session()` and` get_engine()` methods to manage our connections.